### PR TITLE
[Generator] Multipart support

### LINKED ISSUE #36: Support the multipart content type
https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.1.0.md#special-considerations-for-multipart-content

Let's do a proposal first, as there are a few ways the generated code could go here.

### PR DESCRIPTION
**Pull request by czechboy0**
### Motivation

Implement SOAR-0009. Fixes #36.

### Modifications

It's a large diff, and adds some complexity, however I tried to avoid needlessly complicating existing code and opted for duplicating parts of the code specifically for multipart, for easier reasoning.

The implementation very much maps to the changes described in the proposal, which I won't repeat here.

Some notable highlights before you dive in:
- Testing is mainly done through snippet tests, and I added one request and one response operation to make sure things work end to end. Maybe review those first.
- The logic for generating request and response bodies got slightly extended to detect when we're generating multipart content, and branches off to bespoke code specifically for multipart, all in the new `Multipart` directory.
- Made a few related changes, but tried to keep this isolated to multipart as much as reasonable.
- Let me know which parts I should elaborate on, I'm happy to explain in more detail.

### Result

Multipart content now works, as proposed in SOAR-0009.

### Test Plan

Added a bunch of snippet tests and two new file-based reference test operations, including petstore consumer tests to verify that it all works at runtime as well.

### CODE REVIEW

File: Sources/_OpenAPIGeneratorCore/Translator/Content/ContentInspector.swift [hunk 1]
```diff
@@ -162,29 +168,40 @@ extension FileTranslator {
...
+        if contentType.isJSON { return .init(contentType: contentType, schema: contentValue.schema) }
+        if contentType.isUrlEncodedForm { return .init(contentType: contentType, schema: contentValue.schema) }
+        if contentType.isMultipart {
+            guard isRequired else {
+                diagnostics.emit(
+                    .warning(
+                        message:
+                            "Multipart request bodies must always be required, but found an optional one - skipping. Mark as `required: true` to get this body generated.",
+                        context: ["foundIn": foundIn]
+                    )
```
Comment (simonjbeaumont) [inline Sources/_OpenAPIGeneratorCore/Translator/Content/ContentInspector.swift:199]: Can you elaborate on why this is here? Is this something that is convention in OpenAPI docs but compliant with OAS?

File: Sources/_OpenAPIGeneratorCore/Translator/RequestBody/translateRequestBody.swift [hunk 2]
```diff
@@ -45,12 +46,18 @@ extension TypesFileTranslator {
         let contentTypeName = typeName.appending(jsonComponent: "content")
         let contents = requestBody.contents
         for content in contents {
-            if TypeMatcher.isInlinable(content.content.schema) {
+            if TypeMatcher.isInlinable(content.content.schema)
+                || (content.content.contentType.isMultipart
+                    && TypeMatcher.multipartElementTypeReferenceIfReferenceable(
+                        schema: content.content.schema,
+                        encoding: content.content.encoding
+                    ) == nil)
```
Comment (simonjbeaumont) [inline Sources/_OpenAPIGeneratorCore/Translator/RequestBody/translateRequestBody.swift:54]: Could this be chopped up a bit to make it easier to follow the conditions?

File: Sources/_OpenAPIGeneratorCore/Translator/TypeAssignment/TypeMatcher.swift [hunk 3]
```diff
@@ -149,6 +149,21 @@ struct TypeMatcher {
...
+    /// - Parameters:
+    ///   - schema: The schema to try to reference.
+    ///   - encoding: The associated encoding.
+    /// - Returns: A reference if the schema is referenceable, nil otherwise.
+    static func multipartElementTypeReferenceIfReferenceable(
+        schema: UnresolvedSchema?,
+        encoding: OrderedDictionary<String, OpenAPI.Content.Encoding>?
+    ) -> OpenAPI.Reference<JSONSchema>? {
+        // If the schema is a ref AND no encoding is provided, we can reference the type.
+        // Otherwise, we must inline.
```
Comment (simonjbeaumont) [inline Sources/_OpenAPIGeneratorCore/Translator/TypeAssignment/TypeMatcher.swift:162]: This one sounds like it should have some good test coverage; does it? I looked but didn't see it.

**Review by simonjbeaumont — APPROVED** [review]
👏 Great stuff.

File: Sources/_OpenAPIGeneratorCore/Translator/TypeAssignment/TypeMatcher.swift (hunk 3 above)
Comment (czechboy0) [reply Sources/_OpenAPIGeneratorCore/Translator/TypeAssignment/TypeMatcher.swift:162]: Yes, plus also the "is ref" methods.

File: Package.swift [hunk 4]
```diff
@@ -64,7 +64,7 @@ let package = Package(
         // Tests-only: Runtime library linked by generated code, and also
         // helps keep the runtime library new enough to work with the generated
         // code.
-        .package(url: "https://github.com/apple/swift-openapi-runtime", .upToNextMinor(from: "0.3.6")),
+        .package(url: "https://github.com/apple/swift-openapi-runtime", branch: "main"),
```
[the remark this answers is not in the document]
Comment (czechboy0) [reply Package.swift:67]: Discussed offline, no need, the other issues to be fixed need main as well anyway.

File: Sources/_OpenAPIGeneratorCore/Translator/Content/ContentInspector.swift (hunk 1 above)
Comment (czechboy0) [reply Sources/_OpenAPIGeneratorCore/Translator/Content/ContentInspector.swift:199]: It's to catch this case: https://github.com/apple/swift-openapi-generator/pull/369#issuecomment-1825573848

File: Sources/_OpenAPIGeneratorCore/Translator/RequestBody/translateRequestBody.swift (hunk 2 above)
Comment (czechboy0) [reply Sources/_OpenAPIGeneratorCore/Translator/RequestBody/translateRequestBody.swift]: https://github.com/apple/swift-openapi-generator/pull/366/commits/4a802e6c2e4c085fe2fd7d3053544e405d340679

File: Sources/_OpenAPIGeneratorCore/Translator/TypeAssignment/TypeMatcher.swift (hunk 3 above)
Comment (czechboy0) [reply Sources/_OpenAPIGeneratorCore/Translator/TypeAssignment/TypeMatcher.swift:162]: https://github.com/apple/swift-openapi-generator/pull/366/commits/59024dbec6c34a6e0de2982c22055786d64ec62b

**Pull request merged**